From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xucy <xucy0609@gmail.com>
Date: Mon, 01 Jan 2026 00:00:00 +0000
Subject: [PATCH] Config System (consolidated)

Consolidated from 4 original patches:
  - 002-Add-config-for-server-mod-name.patch
  - 003-Add-config-to-disable-async-catchers.patch
  - 004-Add-config-to-avoid-of-folia-supported-check.patch
  - 014-Add-config-for-waypoint-restoration.patch


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -48,14 +48,14 @@
      */
     @Deprecated
     public static void ensureTickThread(final String reason) {
-        if (!isTickThread()) {
+        if (!isTickThread() && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             LOGGER.error("Thread failed main thread check: " + reason + ", context=" + getThreadContext(), new Throwable());
             throw new IllegalStateException(reason);
         }
     }
 
     public static void ensureTickThread(final Level world, final BlockPos pos, final String reason) {
-        if (!isTickThreadFor(world, pos)) {
+        if (!isTickThreadFor(world, pos) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos;
             LOGGER.error(ex, new Throwable());
@@ -64,7 +64,7 @@
     }
 
     public static void ensureTickThread(final Level world, final BlockPos pos, final int blockRadius, final String reason) {
-        if (!isTickThreadFor(world, pos, blockRadius)) {
+        if (!isTickThreadFor(world, pos, blockRadius) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                 reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos + ", block_radius=" + blockRadius;
             LOGGER.error(ex, new Throwable());
@@ -73,7 +73,7 @@
     }
 
     public static void ensureTickThread(final Level world, final ChunkPos pos, final String reason) {
-        if (!isTickThreadFor(world, pos)) {
+        if (!isTickThreadFor(world, pos) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                 reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + pos;
             LOGGER.error(ex, new Throwable());
@@ -82,7 +82,7 @@
     }
 
     public static void ensureTickThread(final Level world, final int chunkX, final int chunkZ, final String reason) {
-        if (!isTickThreadFor(world, chunkX, chunkZ)) {
+        if (!isTickThreadFor(world, chunkX, chunkZ) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                 reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + new ChunkPos(chunkX, chunkZ);
             LOGGER.error(ex, new Throwable());
@@ -91,7 +91,7 @@
     }
 
     public static void ensureTickThread(final Entity entity, final String reason) {
-        if (!isTickThreadFor(entity)) {
+        if (!isTickThreadFor(entity) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                 reason + ", context=" + getThreadContext() + ", entity=" + EntityUtil.dumpEntity(entity);
             LOGGER.error(ex, new Throwable());
@@ -100,7 +100,7 @@
     }
 
     public static void ensureTickThread(final Level world, final AABB aabb, final String reason) {
-        if (!isTickThreadFor(world, aabb)) {
+        if (!isTickThreadFor(world, aabb) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                 reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", aabb=" + aabb;
             LOGGER.error(ex, new Throwable());
@@ -109,7 +109,7 @@
     }
 
     public static void ensureTickThread(final Level world, final double blockX, final double blockZ, final String reason) {
-        if (!isTickThreadFor(world, blockX, blockZ)) {
+        if (!isTickThreadFor(world, blockX, blockZ) && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Luminol
             final String ex = "Thread failed main thread check: " +
                 reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + new Vec3(blockX, 0.0, blockZ);
             LOGGER.error(ex, new Throwable());

diff --git a/src/main/java/com/destroystokyo/paper/network/PaperServerListPingEventImpl.java b/src/main/java/com/destroystokyo/paper/network/PaperServerListPingEventImpl.java
--- a/src/main/java/com/destroystokyo/paper/network/PaperServerListPingEventImpl.java
+++ b/src/main/java/com/destroystokyo/paper/network/PaperServerListPingEventImpl.java
@@ -14,7 +14,7 @@
 
     PaperServerListPingEventImpl(MinecraftServer server, StatusClient client, int protocolVersion, @Nullable CachedServerIcon icon) {
         super(client, server.motd(), server.getPlayerCount(), server.getMaxPlayers(),
-                server.getServerModName() + ' ' + server.getServerVersion(), protocolVersion, icon);
+                fun.bm.mili.lmili.config.modules.misc.ServerModNameConfig.fakeVanilla ? server.getServerVersion() : server.getServerModName() + ' ' + server.getServerVersion(), protocolVersion, icon);//Luminol - Fake vanilla mode
         this.server = server;
     }
 
-- 

diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
--- a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
@@ -25,7 +25,7 @@
     @Override
     public PaperPluginParent build(JarFile file, PaperPluginMeta configuration, Path source) {
         // Folia start - block plugins not marked as supported
-        if (!configuration.isFoliaSupported()) {
+        if (!configuration.isFoliaSupported() && !fun.bm.mili.lmili.config.modules.unsupported.DisableCheckForFoliaSupported.disableForPaper) { // Luminol - disable check for folia supported
             throw new RuntimeException("Could not load plugin '" + configuration.getDisplayName() + "' as it is not marked as supporting Folia!");
         }
         // Folia end - block plugins not marked as supported

diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java b/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
--- a/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
@@ -27,7 +27,7 @@
     @Override
     public SpigotPluginProvider build(JarFile file, PluginDescriptionFile configuration, Path source) throws InvalidDescriptionException {
         // Folia start - block plugins not marked as supported
-        if (!configuration.isFoliaSupported()) {
+        if (!configuration.isFoliaSupported() && !fun.bm.mili.lmili.config.modules.unsupported.DisableCheckForFoliaSupported.disableForPaper) { // Luminol - disable check for folia supported
             throw new RuntimeException("Could not load plugin '" + configuration.getDisplayName() + "' as it is not marked as supporting Folia!");
         }
         // Folia end - block plugins not marked as supported

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1152,7 +1152,7 @@
     }
 
     private void updateWaypoint() {
-        ServerWaypointManager manager = ((ServerLevel) getHandle().level()).getWaypointManager();
+        fun.bm.mili.lmili.utils.FoliaServerWaypointManager manager = ((ServerLevel) getHandle().level()).getWaypointManager(); // Luminol - Restore waypoints
         manager.untrackWaypoint(getHandle());
         manager.trackWaypoint(getHandle());
     }
-- 

diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -393,7 +393,7 @@
         }
 
         // Folia start - block plugins not marked as supported
-        if (!descriptionFile.isFoliaSupported()) {
+        if (!descriptionFile.isFoliaSupported() && !fun.bm.mili.lmili.config.modules.unsupported.DisableCheckForFoliaSupported.disableForPaper) { // Luminol - disable check for folia supported
             throw new InvalidPluginException("Plugin " + descriptionFile.getFullName() + " is not marked as supporting regionised multithreading");
         }
         // Folia end - block plugins not marked as supported
-- 

diff --git a/src/main/java/org/spigotmc/AsyncCatcher.java b/src/main/java/org/spigotmc/AsyncCatcher.java
--- a/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -5,7 +5,7 @@
 public class AsyncCatcher {
 
     public static void catchOp(String reason) {
-        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) { // Paper - chunk system
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThread() && !fun.bm.mili.lmili.config.modules.experiment.DisableAsyncCatcherConfig.enabled) { // Paper - chunk system  // Luminol
             MinecraftServer.LOGGER.error("Thread {} failed main thread check: {}", Thread.currentThread().getName(), reason, new Throwable()); // Paper
             throw new IllegalStateException("Asynchronous " + reason + "!");
         }
-- 

